### PR TITLE
fix $not with $size 0

### PR DESCRIPTION
**Pull request by Yuqi-Du**
This filter should find all the documents with ages array size is not 0, or simply do not have ages array.

```
{
    "find": {
        "filter": {
            "$not": {
                "ages": {
                    "$size": 0
                }
            }
        }
    }
}
```
since we pushed down the $not logic by negating the size number, so, 0 * -1 still -1
FilterableResolver can not know the intention of doing a map_not_equal.

Methodology here, is to add validation first, that $size value should be integer.
Then, since we want to push down the $not info to the FilterableResolver, we need a flag to say this is negating 0.
I choose -1.5 as the flag, since 
- it is a negative number, meaning we have $not logic in it.
- it is not an integer, so will never be user's input, and we can reserve it as a flag indicating negating 0

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java [hunk 1]
```diff
@@ -230,6 +230,11 @@ private void validate(
                   ErrorCode.INVALID_FILTER_EXPRESSION,
                   "$size operator must have integer value >= 0");
             }
+            // Check if the value is an integer by comparing its scale.
+            if (i.stripTrailingZeros().scale() > 0) {
+              throw new JsonApiException(
+                  ErrorCode.INVALID_FILTER_EXPRESSION, "$size operator must have an integer value");
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:236]: Why this is needed, the previous line validation should have taken care of this?

**Review by maheshrajamani — APPROVED** [review]
LGTM.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java [hunk 2]
```diff
@@ -2,9 +2,16 @@
...
+
 /**
  * @param value Literal value to use as RHS operand in the query
  * @param type Literal value type of the RHS operand in the query
  * @param <T> Data type of the object
  */
-public record JsonLiteral<T>(T value, JsonType type) {}
+public record JsonLiteral<T>(T value, JsonType type) {
+
+  // for negating 0, will set BigDecimal -0.5 as flag
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java:14]: Comment and value differ here...

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java [hunk 3]
```diff
@@ -230,6 +230,11 @@ private void validate(
                   ErrorCode.INVALID_FILTER_EXPRESSION,
                   "$size operator must have integer value >= 0");
             }
+            // Check if the value is an integer by comparing its scale.
+            if (i.stripTrailingZeros().scale() > 0) {
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:234]: I don't think we have suitable `NumberUtil` class yet, but in future we might want to create something like `ValueUtil` to collect helper methods like this one here (I did check and this is indeed a good way for checking).

File: src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java [hunk 4]
```diff
@@ -429,6 +429,49 @@ public void mustHandleSize() throws Exception {
           .isEqualTo(expectedResult.getPath());
     }
 
+    @Test
+    public void mustHandleIntegerWithTrailingZeroSize() throws Exception {
+      String json = """
+          {"sizePath" : {"$size": 0.0}}
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java:435]: Is this actually required to be legal? (curious, maybe Mongo allows it?)

**Comment by tatu-at-datastax** [comment]
Ok, aside from smaller questions I had, my main question is this: can we avoid using `BigDecimal` marker, and use either `Boolean` or `String`? Of these, `Boolean.TRUE` might be safest (not the default value of `Boolean.FALSE`, nor same as explicit number value)?

There may be reason this is not doable, so just asking.

**Review by tatu-at-datastax — APPROVED** [review]
LGTM; had some questions but nothing blocking.

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java (hunk 1 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:236]: This is for user inputs that are not Integer.

{
    "find": {
        "filter": {
                "ages": {
                    "$size":  2.43
                }
        }
    }
}

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java (hunk 3 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializer.java:234]: Thanks

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java (hunk 2 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java:14]: fixed, thanks

File: src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java (hunk 4 above)
Comment (Yuqi-Du) [reply src/test/java/io/stargate/sgv2/jsonapi/api/model/command/deserializers/FilterClauseDeserializerTest.java:435]: I checked Mongo Side. They allow it.

**Comment by Yuqi-Du** [comment]
> Ok, aside from smaller questions I had, my main question is this: can we avoid using `BigDecimal` marker, and use either `Boolean` or `String`? Of these, `Boolean.TRUE` might be safest (not the default value of `Boolean.FALSE`, nor same as explicit number value)?
> 
> There may be reason this is not doable, so just asking.

Nice suggestion.
Fixed!

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java [hunk 5]
```diff
@@ -1,7 +1,6 @@
 package io.stargate.sgv2.jsonapi.api.model.command.clause.filter;
 
 // Literal value to use as RHS operand in the query
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/JsonLiteral.java:3]: nit: this comment probably needs to be removed too?

**Review by tatu-at-datastax — APPROVED** [review]

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java [hunk 6]
```diff
@@ -929,6 +929,37 @@ public void withSizeOperatorNoMatch() {
...
     }
 
+    @Test
+    public void withSizeOperatorNotZero() {
+      String json =
+          """
+                  {
+                      "find": {
+                          "filter": {
+                              "$not": {
```
Comment (vkarpov15) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java:939]: The $not syntax that Mongoose supports is `tags: { $not: { $size: 0 } }`, not `{ $not: { tags: { $size: 0 } } }`. Does the `tags: { $not: { $size: 0 } }` syntax work here?

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java [hunk 7]
```diff
@@ -929,6 +929,37 @@ public void withSizeOperatorNoMatch() {
...
+          .contentType(ContentType.JSON)
+          .body(json)
+          .when()
+          .post(CollectionResource.BASE_PATH, namespaceName, collectionName)
+          .then()
+          .statusCode(200)
+          .body("status", is(nullValue()))
+          .body("errors", is(nullValue()))
+          // should have all the documents
+          .body("data.documents", hasSize(6));
```
Comment (vkarpov15) [inline src/test/java/io/stargate/sgv2/jsonapi/api/v1/FindIntegrationTest.java:960]: For completeness, would be great to have a document with `tags: []` and test that's excluded from the output

**Pull request merged**